orchagent/fdborch.cpp

```diff
@@ -118,85 +128,233 @@ void FdbOrch::update(sai_fdb_event_t type, const sai_fdb_entry_t* entry, sai_obj
+            }
+            // dont return, let it delete just to bring SONiC and SAI in sync
+            // return;
+        }
+
+        if (existing_entry->second.type == "static")
+        {
+            if (vlan.m_members.find(update.port.m_alias) == vlan.m_members.end())
+            {
+        	    update.type = "static";
+        	    saved_fdb_entries[update.port.m_alias].push_back({existing_entry->first.mac, vlan.m_vlan_info.vlan_id, "static"});
+            }
+            else
+            {
+                /*port added back to vlan before we receive delete
+                  notification for flush from SAI. Re-add entry to SAI
+                 */ 
+                sai_attribute_t attr;
+                vector<sai_attribute_t> attrs;
+
+                attr.id = SAI_FDB_ENTRY_ATTR_TYPE;
+                attr.value.s32 = SAI_FDB_ENTRY_TYPE_STATIC;
+                attrs.push_back(attr);
+                attr.id = SAI_FDB_ENTRY_ATTR_BRIDGE_PORT_ID;
+                attr.value.oid = bridge_port_id;
+                attrs.push_back(attr);
+                auto status = sai_fdb_api->create_fdb_entry(entry, (uint32_t)attrs.size(), attrs.data());
+                if (status != SAI_STATUS_SUCCESS)
+                {
+                    SWSS_LOG_ERROR("Failed to create FDB %s on %s, rv:%d",
+                        existing_entry->first.mac.to_string().c_str(), update.port.m_alias.c_str(), status);
+                    return;
+                }
+                return;
+            }
+        }
+
         update.add = false;
         storeFdbEntryState(update);
+        update.port.m_fdb_count--;
+        m_portsOrch->setPort(update.port.m_alias, update.port);
+        vlan.m_fdb_count--;
+        m_portsOrch->setPort(vlan.m_alias, vlan);
 
         for (auto observer: m_observers)
         {
             observer->update(SUBJECT_TYPE_FDB_CHANGE, &update);
         }
-
         break;
+    }
+    case SAI_FDB_EVENT_MOVE:
+    {
+        Port port_old;
+        auto existing_entry = m_entries.find(update.entry);
 
+        SWSS_LOG_INFO("Received MOVE event for bvid=%lx mac=%s port=%lx", entry->bv_id, update.entry.mac.to_string().c_str(), bridge_port_id);
+        update.add = true;
+
+        if (!m_portsOrch->getPort(entry->bv_id, vlan))
+        {
+            SWSS_LOG_NOTICE("FdbOrch MOVE notification: Failed to locate vlan port from bv_id 0x%lx", entry->bv_id);
+            return;
+        }
+
+        if (!m_portsOrch->getPortByBridgePortId(bridge_port_id, update.port))
+        {
+            SWSS_LOG_ERROR("FdbOrch MOVE notification: Failed to get port by bridge port ID 0x%lx", bridge_port_id);
+            return;
+        }
+
+        // We should already have such entry
+        if (existing_entry == m_entries.end())
+        {
+             SWSS_LOG_INFO("FdbOrch MOVE notification: mac %s is not found in bv_id 0x%lx",
+                    update.entry.mac.to_string().c_str(), entry->bv_id);
+             break;
+        }
```

Comment:
For mistake-tolerance, even though the old entry is not exist, the new entry still need to be saved in m_entries and stateDB.

Reply:
in that case, we need to treat this as a LEARN case and increment CRM used counter and port fdb count, which may leave us in inconsistent state.

Reply:
For example, a MAC move event and a port destroy event happens at the same time. We must guarantee that the MAC move event is handled before port destroy event, if not, when handle the MAC move, the old port or vlan may be not exist(MAC may be not deleted by port destroy since port is changed). In this case, the old port fdb count may be decreased by port destroy event, and the new port fdb count must be increased by MAC move event.

Reply:
Got it. I have changes code accordingly.